src/abstractitensornetwork.jl

```diff
@@ -721,6 +721,48 @@ function contract_inner(
   return contract(tn; sequence)[]
 end
 
+function contract_with_BP(
+  itn::AbstractITensorNetwork; outputlevel=1, partitioning=group(v -> v, vertices(itn))
+)
+  return contract_with_BP(
+    ComplexF64, itn::AbstractITensorNetwork; outputlevel, partitioning
+  )
+end
+
+function contract_with_BP(
+  T::Type,
+  itn::AbstractITensorNetwork;
+  outputlevel=1,
+  partitioning=group(v -> v, vertices(itn)),
+)
+  @assert isempty(externalinds(itn))
+  bp_cache = BeliefPropagationCache(copy(itn), partitioning)
+
+  bp_cache = update(bp_cache)
+
+  pg = partitioned_itensornetwork(bp_cache)
+
+  if !is_tree(partitioned_graph(pg)) && outputlevel > 0
+    println("Partitioned graph is not a tree, result will be approximate!!")
+  end
+
+  log_numerator, log_denominator = 0, 0
+  for pv in partitionvertices(pg)
+    incoming_mts = incoming_messages(bp_cache, [pv])
```

Comment:
@JoeyT1994 so is this inherently outputting both the messages and the tensors in partition `pv` of the tensor network?

Reply:
Never mind, I see those are being extracted in the next line.